File: src/mod10_design_pattern/lesson13_observer/java/Hogwarts.java
package mod10_design_pattern.lesson13_observer.java;

import java.util.ArrayList;
import java.util.List;

public class Hogwarts implements Subject {

    private List<Observer> observers;
    private String event;

    public Hogwarts() {
        observers = new ArrayList<>();
    }

    @Override
    public void registerObserver(Observer observer) {
        observers.add(observer);
    }

    @Override
    public void removeObserver(Observer observer) {
        observers.remove(observer);
    }

    @Override
    public void notifyObservers() {
        for (Observer observer : observers) {
            observer.update(event);
        }
    }

    public void setEvent(String event) {
        this.event = event;
        notifyObservers();
    }
}

File: src/mod10_design_pattern/lesson13_observer/kotlin/Hogwarts.kt
package mod10_design_pattern.lesson13_observer.kotlin

class Hogwarts : Subject {

    private val observerList = mutableListOf<Observer>()
    private var event: String = ""

    override fun registerObserver(observer: Observer) {
        observerList.add(observer)
    }

    override fun removeObserver(observer: Observer) {
        observerList.remove(observer)
    }

    override fun notifyObservers() {
        observerList.forEach { it.update(event) }
    }

    fun setEvent(event:String){
        this.event=event
        notifyObservers()
    }
}